File: src/main/java/io/github/rcarlosdasilva/cogen/core/Generator.java
package io.github.rcarlosdasilva.cogen.core;

import com.google.common.base.Preconditions;
import com.google.common.collect.Lists;
import io.github.rcarlosdasilva.cogen.config.Configuration;
import io.github.rcarlosdasilva.cogen.core.handler.*;
import lombok.extern.slf4j.Slf4j;

import java.util.List;

/**
 * 代码生成器
 *
 * @author <a href="mailto:rcarlosdasilva@qq.com">Dean Zhao</a>
 */
@Slf4j
public final class Generator {

  private static final Generator INSTANCE = new Generator();

  private static List<Handler> handlers = Lists.newArrayList();

  static {
    handlers.add(new ConfigurationParseHandler());
    handlers.add(new DatabaseReadHandler());
    handlers.add(new FilePrepareHandler());
    handlers.add(new TargetFilterHandler());
    handlers.add(new CodeGenerateHandler());
    handlers.add(new OpenExplorerHandler());
  }

  /**
   * 生成代码
   *
   * @param configuration config
   */
  public static void run(Configuration configuration) {
    Preconditions.checkNotNull(configuration);

    Storage.setConfiguration(configuration);

    log.info("============== Begin ==============");
    handlers.forEach(Handler::handle);
    log.info("============== Done ==============");
  }

}

File: src/main/kotlin/io/github/rcarlosdasilva/cogen/core/Generator.kt
package io.github.rcarlosdasilva.cogen.core

import io.github.rcarlosdasilva.cogen.config.Configuration
import io.github.rcarlosdasilva.cogen.core.handler.*
import mu.KotlinLogging

/**
 * 代码生成器
 *
 * @author [Dean Zhao](mailto:rcarlosdasilva@qq.com)
 */
object Generator {

  private val logger = KotlinLogging.logger {}

  private val handlers: List<Handler> by lazy {
    listOf(
      ConfigurationParseHandler,
      DatabaseReadHandler,
      FilePrepareHandler,
      TargetFilterHandler,
      CodeGenerateHandler,
      OpenExplorerHandler
    )
  }

  /**
   * 生成代码
   *
   * @param configuration config
   */
  fun run(configuration: Configuration) {
    Storage.configuration = configuration

    logger.info("============== Cogen Begin ==============")
    handlers.forEach(Handler::handle)
    logger.info("============== Cogen Done ==============")
  }

}

// TODO 生成枚举类